src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md

```diff
@@ -0,0 +1,123 @@
+description: An overview of smart contract signature generation and verification with EIP-1271. We also walk through Gnosis Safe's EIP-1271 implementation to provide a concrete example for smart contract developers to build on.
+author: nathanhleung
+lang: en
+tags: ["eip-1271", "smart contracts", "verifying", "signing"]
+skill: intermediate
+published: 2023-01-12
+---
+
+The [EIP-1271](https://eips.ethereum.org/EIPS/eip-1271) standard allows smart contracts to verify signatures.
+
+In this tutorial, we give an overview of digital signatures, EIP-1271's background, and [Gnosis Safe](https://gnosis-safe.io/)'s specific implementation of EIP-1271. All together, this can serve as a starting point for implementing EIP-1271 in your own contracts.
+
+## What is a signature?
+
+In this context, a signature (more precisely, a “digital signature”) is a message + some sort of proof that the message came from a specific person/sender/address.
+
+For instance, a digital signature might look like this (usually combined into some sort of data structure):
+
+1. Message: “I want to log in to this website with my Ethereum wallet.”
+2. Signer: My address is `0x000…`
+3. Proof: Here is some proof that I, `0x000…`, actually created this entire message (this is usually something cryptographic).
+
+It’s important to note that a digital signature includes both a “message” and a “signature”.
+
+Why? For instance, if you gave me a contract to sign, and then I cut off the signature page and gave you back only my signatures without the rest of the contract, the contract would not be valid.
+
+In the same way, a digital signature doesn’t mean anything without an associated message!
+
+## Why does EIP-1271 exist?
+
+In order to create a digital signature for use on Ethereum-based blockchains, you generally need a secret private key which no one else knows. This is what makes your signature, yours (no one else can create the same signature without knowledge of the secret key).
+
+Your Ethereum wallet (i.e. your externally-owned account/EOA) has a private key associated with it, and this is the private key that’s typically used when a website or dapp asks you for a signature (e.g. for “Log in with Ethereum”).
+
+An app can [verify a signature](https://docs.alchemy.com/docs/how-to-verify-a-message-signature-on-ethereum) you create using a third-party library like ethers.js [without knowing your private key](https://en.wikipedia.org/wiki/Public-key_cryptography) (because of math) and be confident (again, [because of math](https://crypto.stackexchange.com/questions/48816/discrete-logarithm-problem-of-ecc)) that _you_ were the one that created the signature.
+
+> In fact, because EOA digital signatures are just based on cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
```

Comment:
```suggestion
> In fact, because EOA digital signatures use public-key cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
```